internal/databases/mysql/binlog_server_handler.go

```diff
@@ -0,0 +1,239 @@
+package mysql
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"net"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-mysql-org/go-mysql/mysql"
+	"github.com/go-mysql-org/go-mysql/replication"
+	"github.com/go-mysql-org/go-mysql/server"
+	"github.com/google/uuid"
+	"github.com/wal-g/tracelog"
+	"github.com/wal-g/wal-g/internal"
+	"github.com/wal-g/wal-g/pkg/storages/storage"
+	"github.com/wal-g/wal-g/utility"
+)
+
+var (
+	untilTS      time.Time
+	lastSentGTID string
+)
+
+func addRotateEvent(s *replication.BinlogStreamer, pos mysql.Position) error {
+	// create rotate event
+	rotateBinlogEvent := replication.BinlogEvent{}
+	rotateBinlogEvent.RawData = make([]byte, 4+1+4+4+4+2+8)
+	// generate header
+	// timestamp default 4 bytes
+	binlogEventPos := 4
+	rotateBinlogEvent.RawData[binlogEventPos] = byte(replication.ROTATE_EVENT)
+	binlogEventPos++
+	binary.LittleEndian.PutUint32(rotateBinlogEvent.RawData[binlogEventPos:], 10000)
+	binlogEventPos += 4
+	binary.LittleEndian.PutUint32(rotateBinlogEvent.RawData[binlogEventPos:], uint32(4+1+4+4+4+2+8+len(pos.Name)))
+	binlogEventPos += 4
+	binary.LittleEndian.PutUint32(rotateBinlogEvent.RawData[binlogEventPos:], 0)
+	binlogEventPos += 4
+	binary.LittleEndian.PutUint16(rotateBinlogEvent.RawData[binlogEventPos:], 0)
+	binlogEventPos += 2
+	// set data
+	binary.LittleEndian.PutUint64(rotateBinlogEvent.RawData[binlogEventPos:], uint64(pos.Pos))
+	rotateBinlogEvent.RawData = append(rotateBinlogEvent.RawData, []byte(pos.Name)...)
+
+	return s.AddEventToStreamer(&rotateBinlogEvent)
+}
+
+func waitReplicationIsDone() error {
+	for {
+		cmd, err := internal.GetCommandSetting(internal.MysqlBinlogGetExecutedGTIDCmd)
+		if err != nil {
+			return err
+		}
+		var out bytes.Buffer
+		cmd.Stdout = &out
+		err = cmd.Run()
+		if err != nil {
+			return err
+		}
+		executedGTIDSet := out.String()
+		if strings.Contains(executedGTIDSet, lastSentGTID) {
```

Comment:
lastSentGTID should be read under some kind of lock, like sync.Mutex
Also, we can not use string.Contains here, we should parse lastSentGTID with gomysql.ParseGTIDSet